src/ipc/ipc4/helper.c

```diff
@@ -238,8 +238,14 @@ __cold static int ipc4_create_pipeline_payload_decode(char *data)
 	const struct ipc4_pipeline_ext_payload *hdr =
 		(struct ipc4_pipeline_ext_payload *)data;
 	const struct ipc4_pipeline_ext_object *obj;
-	size_t size =  hdr->payload_words * sizeof(uint32_t);
-	bool last_object = !hdr->data_obj_array;
+	size_t hdr_cache_size = ALIGN_UP(sizeof(*hdr), CONFIG_DCACHE_LINE_SIZE);
+	bool last_object;
+	size_t size;
+
+	sys_cache_data_invd_range((__sparse_force void __sparse_cache *)MAILBOX_HOSTBOX_BASE,
+				  hdr_cache_size);
+	size = hdr->payload_words * sizeof(uint32_t);
```

Comment:
in non-library builds `data` and `hdr` both point to `MAILBOX_HOSTBOX_BASE`, so maybe you want to invalidate `data` instead? That would also make it logical - you invalidate stale caches and then read out new values.